### ISSUE TITLE
Lateral (strafe) movement mechanics

### ISSUE THREAD

**Issue by petervdmeer** [endless-sky/endless-sky#6518]
### Problem Description

When implementing "lateral movement mechanics", then we need to make some choices on how those mechanics will relate to Outfit and ships properties.
We also need to make some decisions on key-bindings, but the "shift-left" and "shift-right" keys seem most popular here (over the option of defining 2 completely separate keys).

### Related Issue Links

#3972 has a reference to strafe movement.
#3715 also mentions lateral movement (but got closed due to many unrelated arguments in the text).

### Desired Solution

Add Outfit properties `"lateral"`, `"lateral energy"` `"lateral heat"`, etc to describe things like the amount of lateral thrust that an engine can generate and the costs of lateral thrusting.

But besides this, also add a `"lateral efficiency"` to ships that indicates how well the chassis of the ship is configured for lateral thrusting (where an efficiency of 0 means that the ship still cannot do lateral thrust, and an efficiency of 1 means that the ship has a 100% efficiency for lateral thrusting.)

This allows for engines that can perform turning thrust, but no lateral thrust (for example if the engine is spinning up and slowing down flywheels for turning the ship). And this also adds some game-play dynamics where some ships are able to perform some (optional) movements while others cannot.

### Alternative Approaches

**Engine Only**
Only add lateral thrust to engines (Outfits) and allow all ships to use lateral thrust. This would make lateral thrust more equally available (or unavailable) to all ships.

**Ship only**
Only add a `"lateral efficiency"` and re-use the "thrust" or "turn" attributes to determine the amount of lateral thrust that a ship can generate. This would remove the need to add new attributes to Outfits/Engines, but I feel that adding the lateral thrust separately to engines would be a good thing.

### Additional Context

Some other thoughts:
- Lateral thrust is nice, but it is not required for a ship to be able to navigate through ES.
- Once we have lateral thrust, then we probably also like to teach the ship-AI to use it.
- We also need to decide how lateral thrust will interact with engine flares, but that is not required for the first implementation. (And can be discussed in a separate issue once the lateral thrust mechanics are implemented.)
- When a ship has strong forward thrust and strong lateral thrust, and is able to use them together, then such ship might be able to achieve a higher maximum speed when using both together than when using either of the trusts alone. (We will need to decide if we want this.)

**Comment by Zitchas** [comment]
Worth noting, as well: VitalChip's "circle the target and stay away from guns & turrets" behavior in Revamp that made Fighters & Drones so much more effective relied at least in part on having strafing. So while it'd be better to have VitalChip's input on this (being the expert on that), it seems reasonable to think that this could enable that same circling behavior (when the ship has sufficient capabilities, anyway. From the Revamp experience, ships need to have significantly higher acceleration than pretty much any of our fighters do currently in order to be possible.)

**Comment by vitalchip** [comment]
I'd add it as a fraction of main thrust. Every ship should have lateral and reverse thrust. It's just basic station keeping. I don't think these things should be niche use.

You need a lot of it, because it's an acceleration, around 50% main thrust. I'd just stick with that for everything. The more thrust the ship has the more you need in lateral to see the same result aanyway, so it works out pretty well.

As far as Ai, it's very useful for reverse landing approach as the ship is often off course slightly but the current process of turning into the velocity vector and using reverse to 'slide' acxross also scrubs off speed and further exacerbates the stopping distance calculations.

Also very useful for convoy freighters, again when following the leader it's a better way to make a small course correction. This (and keeping ships together when moving to land) is another place reverse is really needed. With lateral and reverse a small burst to halt an overshoot and a small burst to centre the ship.. without it a full 180 degree turn to thrust, another 180 to point forward, a small turn and thrust to correct the course followed by the whole lot again as it's overshooting due to the forward burst needed to move to the side..

I don't think I need go into combat use, it's probably the main reason people will want it..

As far as player controls, 2 separate keys, there are many times you need to use turn and lateral togther, using shift to modify the turn keys will prevent that.

I made lateral thrusting able to be controlled fractionally by the Ai, rather than on/off like main thrust, it's often used to make the smallest of course corrections so making it on/off only didn't make a lot of sense.

I did engine points in the usual manner but scaled the flares by the actual amount of thrust, it's a nicer visual but mostly it stops the player seeing a full flare firing rapidly on and off when the Ai is applying 0.01 lateral thrust to correct the course by 5 pixels.. I've since applied the same scaling to the other engine points.

Final note, my experience is slightly different as I have removed drag from my build and all course stability and correction is provided by thrusters. I also turn down the engines as top speed is approached to control top speed, so quite often they are off when lateral thrust is applied, meaning they need to "power share" less often. When they do need to share it's important the thruster gets first bite or ships on marginal power will get stucktrying to correct course and having nothing left to move forward.

**Comment by Amazinite** [comment]
I'd prefer some method which allows us to limit which ships can use lateral thrust, or at the very least some method which allows some ships to have lots of lateral thrust while others have a miniscule amount meant only for minor adjustments while station keeping, as Vital suggested. Reason being that lateral thrust likely won't be a thing until after #5271/#5772/#5779/#6481 are resolved, so we need to think of implementing lateral thrust into a system where mobility across ship sizes is more balanced.

Although #5271 refers to speed balance being out of whack, the underlying issue is really with mobility in general; larger ships are simply too mobile compared to smaller ships. Mobility currently includes top speed, acceleration, and turning, and to a lesser extent reverse thrust and afterburners, and the addition of lateral thrust would then become another mobility factor to keep in check. I believe that larger ships in general should be less mobile than smaller ships, so I'd rather not have massive heavy warships laterally thrusting around in combat, instead either relegating lateral thrust entirely to drones, fighters, interceptors, and maybe some light warships, or simply only giving lateral thrust significant enough to be useful in combat to these ships, which would give a massive mobility boost to the ships which rely on their mobility the most during combat while leaving larger ships lower in terms of mobility.

**Comment by Zitchas** [comment]
I definitely like the sounds of lateral thrust as VitalChip describes, but I also feel that Amazinite's breakdown of which ships have large amounts of it is the best route forward: While every ship should have some lateral thrust for station keeping and manoeuvring purposes, significant amounts of it (aka the levels that make it a strong factor in combat) should be reserved for highly agile ships.

While there can and should be exceptions, as a general rule I think the only ships that should have strong lateral movement is the interceptor, fighter, and drone categories. Light warships probably also have combat-effective strafing, but as a general rule medium and heavy warships' abilities to strafe should probably only be relevant in combat for things like, say, sidestepping extreme range slow moving artillery fire that they can see coming from across the system, or stuff like that.

**Comment by pepoluan** [comment]
> I'd prefer some method which allows us to limit which ships can use lateral thrust

In real life, a rear-mounted engine will tend to rotate the ship around the center of gravity.

Sure, this can be countered by steering.

But if we want, we can limit lateral movements only to ships equpped with reversers (or using Remnant engines, as they have built-in reversers).

**Comment by Zitchas** [comment]
That's definitely an option. Overall, I'll be happy with any solution that allows us some granularity. As in, lets us specify on per-hull basis, which ships are setup to to handle significant amounts of strafing thrust and which ones aren't. Ideally, with values so we can specify that one ship can do it a lot, another can do it somewhat, and another can barely do it at all.

**Comment by TheGiraffe3** [comment]
@petervdmeer will this happen?
It looks like something that would be very fun to have.

**Comment by petervdmeer** [comment]
> @petervdmeer will this happen? It looks like something that would be very fun to have.

I read in the thread here that VitalChip already implemented AI support for lateral movements somewhere. So I guess better to ask VitalChip than me (since I only created this ticket and did nothing else yet for lateral movement).

**Comment by TheGiraffe3** [comment]
Well, @vitalchip hasn't been active since February of 2023, so I'm not sure they will be able to do it.

**Comment by petervdmeer** [comment]
> Well, @vitalchip hasn't been active since February of 2023, so I'm not sure they will be able to do it.

That's a pity, but I read in this thread that Zitchas also knows where to find the relevant code (since it got embedded in Revamp). @Zitchas Do you indeed have a link to the code that implements lateral movement?

**Comment by Zitchas** [comment]
That would be in here:
https://github.com/endless-sky/Endless-Sky-Delta/pull/10

and here:
https://github.com/endless-sky/Endless-Sky-Delta/pull/14

PR 10 has the meat of what you are looking for. 14 just adds a fallback so that if lateral thrust isn't specified for that ship, it calculates a value for small ships, and uses 0.1 of regular thrust for big ships. This formula was subsequently adjusted after the mass increase.

Note that the original branch for PR 10 still exists. I didn't delete it, so you are welcome to pull directly from it.

**Comment by TheGiraffe3** [comment]
Would there be interest in pulling Delta's version of lateral thrust out, just so that it's in the game?
Note that the values described in the original post don't exist yet. 
But it is possible to have lateral thrust anyway.
